### PR TITLE
Add pngcrush optimizer

### PR DESCRIPTION
**Pull request by bjoernhaeuser**
Add a pngcrush optimizer for optimizing pngs. I dont know what the best options for pngcrush are, tested them and thought that they are fine.

### CODE REVIEW

**Comment by bjoernhaeuser** [comment]
I did not add any tests, because the optimizers are generally not tested. Will add tests if needed.

**Comment by bjoernhaeuser** [comment]
I have added tests.

File: tests/test_optimizers.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from thumbor.optimizers.pngcrush import Optimizer as PngcrushOptimizer
+
+fixtures_folder = join(abspath(dirname(__file__)), 'fixtures')
+
+
+class PngcrushOptimizerTest(unittest.TestCase):
+    def setUp(self):
+        self.pngcrush_path = '/usr/local/bin/pngcrush'
+        if not (os.path.isfile(self.pngcrush_path) and os.access(self.pngcrush_path, os.X_OK)):
+            raise unittest.SkipTest("Unable to locate pngcrush at {}".format(self.pngcrush_path))
```
Comment (bjoernhaeuser) [inline tests/test_optimizers.py]: afaics this won't work for python 2.6. Will try to test.

File: tests/test_optimizers.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    def get_context(self):
+        conf = {'config': Config().load(path=None, defaults={'PNGCRUSH_PATH': self.pngcrush_path})}
```
Comment (bjoernhaeuser) [inline tests/test_optimizers.py]: there must be a better way to construct this config-object, right?

File: tests/test_optimizers.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        self.assertFalse(optimizer.should_run('jpeg', None))
+
+    def test_pngcrush_should_not_run_for_png(self):
+        optimizer = PngcrushOptimizer(self.get_context())
+        self.assertTrue(optimizer.should_run('png', None))
+
+    def test_pngcrush_should_optimize_png(self):
+        optimizer = PngcrushOptimizer(self.get_context())
+        temp = tempfile.NamedTemporaryFile()
+        optimizer.optimize(None, fixtures_folder + '/img/fixture2.png', temp.name)
```
Comment (bjoernhaeuser) [inline tests/test_optimizers.py]: I needed a new png fixture, `fixture1.png` is a jpeg file under the hood.

**Comment by kevin1024** [comment]
Just as a heads up, I already did a pull request for this, and it was rejected in favor of creating a separate project: https://github.com/thumbor/thumbor/pull/387

**Comment by bjoernhaeuser** [comment]
@kevin1024 Oh. Well, so, we have to start a thumbor-optimizers repo somewhere?

**Comment by guilhermef** [comment]
@bjoernhaeuser We've dropped python 2.6 support a while ago.
Are you still using the 2.6 ?

**Comment by bjoernhaeuser** [comment]
@guilhermef oh great, no I am using 2.7, but the doc says that >= 2.6 is needed. Will fix that.

File: tests/test_optimizers.py (hunk 1 above)
Comment (bjoernhaeuser) [inline tests/test_optimizers.py]: 2.6 not needed anymore :)

**Comment by bjoernhaeuser** [comment]
@guilhermef Okay, removed the mentiong of 2.6 in the `CONTRIBUTING.md`. Is this than mergeable or do we want to extract a seperate project?

**Comment by guilhermef** [comment]
@heynemann @dhardy92 @cezarsa what do you guys think ?
Should we create a new repo for optimizers, or keep them on the thumbor repo for now to avoid breaking changes ?

**Comment by bjoernhaeuser** [comment]
Does anyone knows why the tests are failing now? :-(

**Comment by dhardy92** [comment]
according to me thumbor core should only incluse what is necessary to run it naively just after pip install . what have external dependency should be provided as a plugin.
But I am a kind of extremist 😀

**Comment by bjoernhaeuser** [comment]
@dhardy92 I am thinking the same. What's next than?

**Comment by guilhermef** [comment]
I'll create a new project for that tomorrow, ok?

**Comment by bjoernhaeuser** [comment]
Great!

**Comment by guilhermef** [comment]
@bjoernhaeuser I've created the https://github.com/thumbor/thumbor-contrib project to store new optimizers.
Do you mind making this PR there ?

**Comment by bjoernhaeuser** [comment]
I will! Thanks!

**Comment by guilhermef** [comment]
@bjoernhaeuser let me change the project's name, I was talking to @heynemann and he gave a better idea.

**Comment by guilhermef** [comment]
@bjoernhaeuser now the project is https://github.com/thumbor/thumbor-png-optimizers

**Comment by efuquen** [comment]
@guilhermef I'm not sure I understand why the repo was renamed to thumbor-png-optimizers?  Are png optimizers the *only* type of optimizers that will be separate from the core?  Should we have a thumbor-jpeg-optimizers and a thumbor-gif-optimizers repo?  And then what about other plugin types, like i.e. the loaders?

To me its confusing to separate only certain things on such a fine grained level and yet have other things be baked directly into core.  What is the future here, is the plan to have many many repos for different types of plugins?

**Comment by dhardy92** [comment]
I agree with @efuquen maybe some thumbor-plugins repository would be created that can be seen as a plugin catalog.
And starting moving some part of code (mongo and redis storage, optimizers, error handlers, some filters ?) currently in thumbor core as plugins ?

**Comment by guilhermef** [comment]
I thought it would be better to have small components projects.
And When you set your thumbor it just install what you need, not a big plugins project.

We were imagining a requirements.txt like this:
thumbor-http-loader
thumbor-png-optimizers
thumbor-redis-storage

**Comment by dhardy92** [comment]
Yes, I agree that. 
But I thougth that a big git repo with all (with thumbor support) plugins would easier to manage and expose, update (adding a new plugin is just a PR away)
eventually group them by type (storage, optimizer, filter, loader) or dependencies (mongodb service, exec, redis)
Maybe pyvows tests can have common code base (ex : all storages should pass a minimal set of tests)

```
thumbor-plugins/
  README.md
  thumbor-redis-storage/
    setup.py
    ...
  thumbor-png-optimizers/
    setup.py
    ...
```
( according to me thumbor-http-loader shoudl stay in thumbor core )

**Comment by guilhermef** [comment]
To make it simple, let's keep all the contributed filters and optimizers on this project with only one setup.py.
And for mongo and redis storage I still think we should put into different projects.
It would be easier to test them and we'll have their external dependencies running on each one.
What do you guys think ?

**Comment by masom** [comment]
@guilhermef :+1: for keeping mongo, redis, etc. in separate repos. Random unneeded dependencies makes it harder to keep everything in sync all the time.

How would these module be able to tell `thumbor-config`/`Config` their default options?

**Comment by dhardy92** [comment]
@masom I suppose just as I do, in __init__.py, with my hbase storage plugin : https://github.com/dhardy92/thumbor_hbase/blob/master/thumbor_hbase/__init__.py

**Comment by masom** [comment]
@dhardy92 Sweet.

**Comment by bjoernhaeuser** [comment]
Closed in favor of thumbor/thumbor-plugins#1

**Pull request closed without merging**